fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/ComparisonRange.java

```diff
@@ -0,0 +1,166 @@
+        this.equalityComparison = equalityComparison;
+        this.inequalityComparisons = null;
+    }
+
+    private ComparisonRange(@Nonnull List<Comparisons.Comparison> inequalityComparisons) {
+        this.equalityComparison = null;
+        this.inequalityComparisons = inequalityComparisons;
+    }
+
+    public boolean isEmpty() {
+        return equalityComparison == null && inequalityComparisons == null;
+    }
+
+    public boolean isEquality() {
+        return equalityComparison != null && inequalityComparisons == null;
+    }
+
+    public boolean isInequality() {
+        return equalityComparison == null && inequalityComparisons != null;
+    }
+
+    @Nonnull
+    public Comparisons.Comparison getEqualityComparison() {
+        if (equalityComparison == null) {
+            throw new RecordCoreException("tried to get non-existent equality comparison from ComparisonRange");
+        }
+        return equalityComparison;
+    }
+
+    @Nullable
+    public List<Comparisons.Comparison> getInequalityComparisons() {
+        if (inequalityComparisons == null) {
+            throw new RecordCoreException("tried to get non-existent inequality comparisons from ComparisonRange");
+        }
+        return inequalityComparisons;
+    }
+
+    @Nonnull
+    public Optional<ComparisonRange> tryToAdd(@Nonnull Comparisons.Comparison comparison) {
+        // TODO This isn't actually right becuase of how ranges work...
+        // It appears to be compatible with the AndWithThenPlanner
+        if (isEmpty()) {
+            return from(comparison);
+        } else if (isEquality() && getEqualityComparison().equals(comparison)) {
+            return Optional.of(this);
+        } else if (isInequality() &&
+                ScanComparisons.getComparisonType(comparison).equals(ScanComparisons.ComparisonType.INEQUALITY)) {
+            return Optional.of(new ComparisonRange(ImmutableList.<Comparisons.Comparison>builder()
+                    .addAll(inequalityComparisons)
+                    .add(comparison)
+                    .build()));
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public String toString() {
+        if (isEquality()) {
+            return getEqualityComparison().toString();
+        } else if (isInequality()) {
+            return inequalityComparisons.stream().map(Comparisons.Comparison::toString)
+                            .collect(Collectors.joining(" && ", "[", "]"));
+        } else {
+            return "[]";
+        }
+    }
+
+    @Nonnull
+    public static Optional<ComparisonRange> from(@Nonnull Comparisons.Comparison comparison) {
+        switch (ScanComparisons.getComparisonType(comparison)) {
+            case EQUALITY:
+                return Optional.of(new ComparisonRange(comparison));
+            case INEQUALITY:
+                return Optional.of(new ComparisonRange(Collections.singletonList(comparison)));
+            case NONE:
+                return Optional.empty();
```

Comment:
Maybe the imagination isn't that this is where canonicalization takes place, but `NOT NULL` is the same as `> NULL` because we know it's the smallest possible value and `>= 3` and `NOT_EQUALS 3` is `> 3`.

Reply:
Since I'm farming out the classification to `ScanComparisons.getComparisonType()`, it looks like the first case is handled since `NOT_NULL` has type `ComparisonType.INEQUALITY` (for this reason, I presume). I think that we should normalize the second case but doing so is tricky and not an immediate priority, I think. I'll add a TODO comment.